In [1]:
import pandas as pd
import numpy as np
import tsfresh as tsf
from tsfresh import extract_features, select_features
from tsfresh.utilities.dataframe_functions import impute

In [2]:
data_path = '/Users/chiang/数据/天池-心跳信号分类预测/'
data_train = pd.read_csv(data_path + 'train.csv')
data_test_A = pd.read_csv(data_path + 'testA.csv')
print("Training data shape:", data_train.shape)
print("Testing data shape:", data_test_A.shape)

Training data shape: (100000, 3)
Testing data shape: (20000, 2)


In [3]:
# 对心电特征进行行转列处理，同时为每个心电信号加入时间步特征time
train_heartbeat_df = data_train["heartbeat_signals"].str.split(",", expand=True).stack()
train_heartbeat_df = train_heartbeat_df.reset_index()
train_heartbeat_df = train_heartbeat_df.set_index("level_0")
train_heartbeat_df.index.name = None
train_heartbeat_df.rename(columns={"level_1":"time", 0:"heartbeat_signals"}, inplace=True)
train_heartbeat_df["heartbeat_signals"] = train_heartbeat_df["heartbeat_signals"].astype(float)

train_heartbeat_df

time  heartbeat_signals
0         0           0.991230
0         1           0.943533
0         2           0.764677
0         3           0.618571
0         4           0.379632
...     ...                ...
99999   200           0.000000
99999   201           0.000000
99999   202           0.000000
99999   203           0.000000
99999   204           0.000000

[20500000 rows x 2 columns]

In [4]:
# 将处理后的心电特征加入到训练数据中，同时将训练数据label列单独存储
data_train_label = data_train["label"]
data_train = data_train.drop("label", axis=1)
data_train = data_train.drop("heartbeat_signals", axis=1)
data_train = data_train.join(train_heartbeat_df)

data_train

id  time  heartbeat_signals
0          0     0           0.991230
0          0     1           0.943533
0          0     2           0.764677
0          0     3           0.618571
0          0     4           0.379632
...      ...   ...                ...
99999  99999   200           0.000000
99999  99999   201           0.000000
99999  99999   202           0.000000
99999  99999   203           0.000000
99999  99999   204           0.000000

[20500000 rows x 3 columns]

In [5]:
data_train[data_train["id"]==1]

id  time  heartbeat_signals
1    1     0           0.971482
1    1     1           0.928969
1    1     2           0.572933
1    1     3           0.178457
1    1     4           0.122962
..  ..   ...                ...
1    1   200           0.000000
1    1   201           0.000000
1    1   202           0.000000
1    1   203           0.000000
1    1   204           0.000000

[205 rows x 3 columns]

每个样本的心电特征都由205个时间步的心电信号组成。



## 使用 tsfresh 进行时间序列特征处理
特征抽取 **Tsfresh（TimeSeries Fresh）**是一个Python第三方工具包。 它可以自动计算大量的时间序列数据的特征。此外，该包还包含了特征重要性评估、特征选择的方法，因此，不管是基于时序数据的分类问题还是回归问题，tsfresh都会是特征提取一个不错的选择。

In [6]:
from tsfresh import extract_features

# 特征提取
train_features = extract_features(data_train, column_id='id', column_sort='time')
train_features

Feature Extraction: 100%|██████████| 10/10 [1:56:34<00:00, 699.48s/it] 


heartbeat_signals__variance_larger_than_standard_deviation  \
0                                                    0.0            
1                                                    0.0            
2                                                    0.0            
3                                                    0.0            
4                                                    0.0            
...                                                  ...            
99995                                                0.0            
99996                                                0.0            
99997                                                0.0            
99998                                                0.0            
99999                                                0.0            

       heartbeat_signals__has_duplicate_max  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
4                                       0.0   
...                                     ...   
99995                                   0.0   
99996                                   0.0   
99997                                   0.0   
99998                                   0.0   
99999                                   0.0   

       heartbeat_signals__has_duplicate_min  heartbeat_signals__has_duplicate  \
0                                       1.0                               1.0   
1                                       1.0                               1.0   
2                                       1.0                               1.0   
3                                       1.0                               1.0   
4                                       1.0                               1.0   
...                                     ...                               ...   
99995                                   1.0                               1.0   
99996                                   1.0                               1.0   
99997                                   1.0                               1.0   
99998                                   1.0                               1.0   
99999                                   1.0                               1.0   

       heartbeat_signals__sum_values  heartbeat_signals__abs_energy  \
0                          38.927945                      18.216197   
1                          19.445634                       7.705092   
2                          21.192974                       9.140423   
3                          42.113066                      15.757623   
4                          69.756786                      51.229616   
...                              ...                            ...   
99995                      63.323449                      28.742238   
99996                      69.657534                      31.866323   
99997                      40.897057                      16.412857   
99998                      42.333303                      14.281281   
99999                      53.290117                      21.637471   

       heartbeat_signals__mean_abs_change  heartbeat_signals__mean_change  \
0                                0.019894                       -0.004859   
1                                0.019952                       -0.004762   
2                                0.009863                       -0.004902   
3                                0.018743                       -0.004783   
4                                0.014514                        0.000000   
...                                   ...                             ...   
99995                            0.023588                       -0.004902   
99996                            0.017373                       -0.004543   
99997                            0.019470                       -0.004538   
99998            

上一步跑了两个小时。。。
### 特征选择
 train_features中包含了heartbeat_signals的787种常见的时间序列特征（所有这些特征的解释可以去看官方文档），这其中有的特征可能为NaN值（产生原因为当前数据不支持此类特征的计算），使用以下方式去除NaN值。

In [7]:
from tsfresh.utilities.dataframe_functions import impute

# 去除抽取特征中的NaN值
impute(train_features)


/opt/anaconda3/envs/nlp/lib/python3.7/site-packages/tsfresh/utilities/dataframe_functions.py:172: RuntimeWarning: The columns ['heartbeat_signals__query_similarity_count__query_None__threshold_0.0'] did not have any finite values. Filling with zeros.
  df.iloc[:, np.where(is_col_non_finite)[0]].columns.values), RuntimeWarning)


heartbeat_signals__variance_larger_than_standard_deviation  \
0                                                    0.0            
1                                                    0.0            
2                                                    0.0            
3                                                    0.0            
4                                                    0.0            
...                                                  ...            
99995                                                0.0            
99996                                                0.0            
99997                                                0.0            
99998                                                0.0            
99999                                                0.0            

       heartbeat_signals__has_duplicate_max  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
4                                       0.0   
...                                     ...   
99995                                   0.0   
99996                                   0.0   
99997                                   0.0   
99998                                   0.0   
99999                                   0.0   

       heartbeat_signals__has_duplicate_min  heartbeat_signals__has_duplicate  \
0                                       1.0                               1.0   
1                                       1.0                               1.0   
2                                       1.0                               1.0   
3                                       1.0                               1.0   
4                                       1.0                               1.0   
...                                     ...                               ...   
99995                                   1.0                               1.0   
99996                                   1.0                               1.0   
99997                                   1.0                               1.0   
99998                                   1.0                               1.0   
99999                                   1.0                               1.0   

       heartbeat_signals__sum_values  heartbeat_signals__abs_energy  \
0                          38.927945                      18.216197   
1                          19.445634                       7.705092   
2                          21.192974                       9.140423   
3                          42.113066                      15.757623   
4                          69.756786                      51.229616   
...                              ...                            ...   
99995                      63.323449                      28.742238   
99996                      69.657534                      31.866323   
99997                      40.897057                      16.412857   
99998                      42.333303                      14.281281   
99999                      53.290117                      21.637471   

       heartbeat_signals__mean_abs_change  heartbeat_signals__mean_change  \
0                                0.019894                       -0.004859   
1                                0.019952                       -0.004762   
2                                0.009863                       -0.004902   
3                                0.018743                       -0.004783   
4                                0.014514                        0.000000   
...                                   ...                             ...   
99995                            0.023588                       -0.004902   
99996                            0.017373                       -0.004543   
99997                            0.019470                       -0.004538   
99998            

接下来，按照特征和响应变量之间的相关性进行特征选择，这一过程包含两步：首先单独计算每个特征和响应变量之间的相关性，然后利用Benjamini-Yekutieli procedure 进行特征选择，决定哪些特征可以被保留。

In [8]:
from tsfresh import select_features

# 按照特征和数据label之间的相关性进行特征选择
train_features_filtered = select_features(train_features, data_train_label)

train_features_filtered

heartbeat_signals__sum_values  \
0                          38.927945   
1                          19.445634   
2                          21.192974   
3                          42.113066   
4                          69.756786   
...                              ...   
99995                      63.323449   
99996                      69.657534   
99997                      40.897057   
99998                      42.333303   
99999                      53.290117   

       heartbeat_signals__fft_coefficient__attr_"abs"__coeff_38  \
0                                               0.660949          
1                                               1.718217          
2                                               1.814281          
3                                               2.109550          
4                                               0.194549          
...                                                  ...          
99995                                           0.840651          
99996                                           1.557787          
99997                                           0.469758          
99998                                           0.992948          
99999                                           1.624625          

       heartbeat_signals__fft_coefficient__attr_"abs"__coeff_37  \
0                                               1.090709          
1                                               1.280923          
2                                               1.619051          
3                                               0.619634          
4                                               0.348882          
...                                                  ...          
99995                                           1.186210          
99996                                           1.393960          
99997                                           1.000355          
99998                                           1.354894          
99999                                           1.739088          

       heartbeat_signals__fft_coefficient__attr_"abs"__coeff_36  \
0                                               0.848728          
1                                               1.850706          
2                                               1.215343          
3                                               2.366413          
4                                               0.092119          
...                                                  ...          
99995                                           1.396236          
99996                                           0.989147          
99997                                           0.706395          
99998                                           2.238589          
99999                                           2.936555          

       heartbeat_signals__fft_coefficient__attr_"abs"__coeff_35  \
0                                               1.168685          
1                                               1.460752          
2                                               1.787166          
3                                               2.071539          
4                                               0.653924          
...                                                  ...          
99995                                           0.417221          
99996                                           1.611333          
99997                                           1.190514          
99998                                           1.237608          
99999                                           0.154759          

       heartbeat_signals__fft_coefficient__attr_"abs"__coeff_34  \
0                                               0.982133          
1                                               1.924501          
2                                               2.146987          
3                                     

可以看到经过特征选择，留下了707个特征。